src/main/scala/chisel3/util/MemInterface.scala

```diff
@@ -0,0 +1,337 @@
+package chisel3.util
+
+import chisel3._
+
+import chisel3.internal.Builder
+import chisel3.experimental.SourceInfo
+import chisel3.internal.sourceinfo.{MemTransform, SourceInfoTransform}
+import scala.language.reflectiveCalls
+import scala.language.experimental.macros
+
+/** A bundle of signals representing a read memory port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  */
+class MemRdPortInterface[T <: Data](tpe: T, addrWidth: Int) extends Bundle {
+  val addr = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val readValue = Output(tpe)
+}
+
+/** A bundle of signals representing a write memory port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  * @param withMask Whether this read/write port interface should have an optional mask.
+  *
+  * @note `withMask` is only valid if tpe is a Vec; if this is not the case no mask will be initialized
+  *       regardless of the value of `withMask`.
+  */
+class MemWrPortInterface[T <: Data](tpe: T, addrWidth: Int, withMask: Boolean) extends Bundle {
+  val maskSize = tpe match {
+    case vec: Vec[_] => vec.size
+    case _ => 0
+  }
+
+  val addr = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val writeValue = Input(tpe)
+  val mask: Option[Vec[Bool]] = if (withMask && maskSize > 0) Some(Input(Vec(maskSize, Bool()))) else None
+}
```

Comment:
Related to discussion above

```suggestion
class MemWritePort[T <: Data](tpe: T, addrWidth: Int, masked: Boolean) extends Bundle {
  val addr = Input(UInt(addrWidth.W))
  val enable = Input(Bool())
  val data = Input(tpe)
  val mask: Option[Vec[Bool]] = if (masked) {
    val maskSize = tpe match {
      case vec: Vec[_] => vec.size
      case bad => require(false, s"Write ports can only be masked with Vec data types! Got $bad")
    }
    Some(Input(Vec(maskSize, Bool())))
  } else {
    None
  }
}
```
It's also not useful to have `maskSize` as a field when there is no mask, so I think it's better to hide it as a local variable.